### PR TITLE
[discovery-proxy] compare DNS labels in a case-insensitive manner

### PR DESCRIPTION
**Pull request by simonlingoogle**

### CODE REVIEW

File: src/utils/string_utils.cpp [hunk 1]
```diff
@@ -0,0 +1,54 @@
...
+bool EqualCaseInsensitive(const std::string &aString1, const std::string &aString2)
+{
+    return ToLowercase(aString1) == ToLowercase(aString2);
+}
+
+std::string ToLowercase(const std::string &aString)
+{
+    std::string ret = aString;
+
+    std::transform(ret.begin(), ret.end(), ret.begin(), [](unsigned char c) { return std::tolower(c); });
```
Comment (superwhd) [inline src/utils/string_utils.cpp:48]:
```suggestion
    std::transform(ret.begin(), ret.end(), ret.begin(), [](char c) { return std::tolower(c); });
```
`std::string` is `std::basic_string<char>`.

File: src/utils/string_utils.hpp [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace otbr {
+
+namespace StringUtils {
+
+/**
+ * This method compares two strings in a case-insensitive manner.
+ *
+ * @param[in] aString1 The first string.
+ * @param[in] aString1 The second string.
```
Comment (superwhd) [inline src/utils/string_utils.hpp:50]:
```suggestion
 * @param[in] aString2 The second string.
```

**Review by superwhd — APPROVED** [review]
LGTM with some nits

File: src/utils/string_utils.hpp [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+ *
+ */
+bool EqualCaseInsensitive(const std::string &aString1, const std::string &aString2);
+
+/**
+ * This method converts a given string to lowercase.
+ *
+ * @param[in] aString The string to convert.
+ *
+ * @returns  A copy of `aString` with all letters converted to lowercase.
```
Comment (jwhui) [inline src/utils/string_utils.hpp:62]:
```suggestion
 * @returns  A copy of @p aString with all letters converted to lowercase.
```

File: src/utils/string_utils.hpp [hunk 4]
```diff
@@ -0,0 +1,71 @@
...
+
+#include <string.h>
+#include <string>
+
+namespace otbr {
+
+namespace StringUtils {
+
+/**
+ * This method compares two strings in a case-insensitive manner.
```
Comment (jwhui) [inline src/utils/string_utils.hpp:47]:
```suggestion
 * This function compares two strings in a case-insensitive manner.
```

File: src/utils/string_utils.hpp [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+ * @param[in] aString1 The first string.
+ * @param[in] aString2 The second string.
+ *
+ * @returns  Whether the two strings are euqal in a case-insensitive manner.
+ *
+ */
+bool EqualCaseInsensitive(const std::string &aString1, const std::string &aString2);
+
+/**
+ * This method converts a given string to lowercase.
```
Comment (jwhui) [inline src/utils/string_utils.hpp:58]:
```suggestion
 * This function converts a given string to lowercase.
```

File: src/utils/string_utils.hpp [hunk 6]
```diff
@@ -0,0 +1,71 @@
...
+ *  SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+ *  INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+ *  CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+ *  ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+
+/**
+ * @file
+ * This file includes definition for strcpy_safe().
```
Comment (jwhui) [inline src/utils/string_utils.hpp:31]: Update file description.

File: src/utils/string_utils.hpp [hunk 7]
```diff
@@ -0,0 +1,71 @@
...
+
+namespace StringUtils {
+
+/**
+ * This method compares two strings in a case-insensitive manner.
+ *
+ * @param[in] aString1 The first string.
+ * @param[in] aString2 The second string.
+ *
+ * @returns  Whether the two strings are euqal in a case-insensitive manner.
```
Comment (jwhui) [inline src/utils/string_utils.hpp:52]:
```suggestion
 * @returns  Whether the two strings are equal in a case-insensitive manner.
```

**Review by jwhui — APPROVED** [review]

**Pull request merged**